### PR TITLE
add new capabilities for language servers

### PR DESCRIPTION
**Pull request by ggilmore**

### CODE REVIEW

**Comment by laughedelic** [comment]
These capabilities don't looks like part of the LSP. Where does it come from?

**Comment by felixfbecker** [comment]
@laughedelic "Automatic dependency management" and "No arbitrary code execution" are not specific methods of the protocol, but general traits of a language server that are useful to know (especially when wanting to run a language server as a service).
"Files extension" and "Symbol Descriptor Extension" are two extensions to LSP we specced out a long time ago to support the use case of running language servers in isolated containers and to support cross-repository code intelligence. They are implemented by a handful of language servers.
The removed "Streaming extension" was another extension, but it proved to be not as useful as we imagined it to be.

File: index.html [hunk 1]
```diff
@@ -781,11 +916,35 @@ <h4>Language servers</h4>
...
+					<strong>No arbitrary code execution</strong>: Language servers that support this feature <em>don't</em> execute arbitrary code (some language servers do 
+					this when running build scripts, analyzing the project, etc.). 
+			</li>
+			<li id="filesExtensionFootnote">
+					<strong>Files extension</strong>: Language servers that support this feature are able to operate without sharing a physical 
+					filesystem with the client. 
+					<a href="https://github.com/sourcegraph/language-server-protocol/blob/master/extension-files.md">See this link for more information.</a> 
+			</li>
+			<li id="symbolDescriptorFootnote">
+					<strong>Symbol Descriptor Extension</strong>: Language servers that support this feature are able to interact with <code>SymbolDescriptor</code>s.
```
[the remark this answers is not in the document]
Comment (ggilmore) [reply index.html]: Probably `operate` is a better word, but I'll use your alternative

**Comment by ggilmore** [comment]
Before: 
screen shot 2018-05-09 at 11 59 28-fullpage

After:
screen shot 2018-05-09 at 11 59 08-fullpage

@felixfbecker PTAL

File: index.html [hunk 2]
```diff
@@ -562,8 +643,11 @@ <h4>Language servers</h4>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
+				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
 				<td class="danger"></td>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
+				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
+				<td class="danger"></td>
```
Comment (felixfbecker) [inline index.html]: PHP supports the symbol descriptor extension.

File: index.html [hunk 3]
```diff
@@ -562,8 +643,11 @@ <h4>Language servers</h4>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
+				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
```
Comment (felixfbecker) [inline index.html]: PHP does not support dependency installation.

File: index.html [hunk 4]
```diff
@@ -406,8 +448,11 @@ <h4>Language servers</h4>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
 				<td class="success"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span></td>
+				<td class="danger"></td>
+				<td class="danger"></td>
```
Comment (felixfbecker) [inline index.html]: Go supports the files extension.

File: index.html [hunk 5]
```diff
@@ -781,11 +916,35 @@ <h4>Language servers</h4>
...
 		</table>
 
+		<h5>Notes</h5>
+		<ol>
+			<li id="dependencyFootnote">
+				<strong>Automatic dependency management</strong>: Langugage servers that support this feature are able to resolve / install a project's 
+				3rd-party depdencies without the need for a user to manually intervene.
+			</li>
+			<li id="arbitraryExecutionFootnote">
+					<strong>No arbitrary code execution</strong>: Language servers that support this feature <em>don't</em> execute arbitrary code (some language servers do
```
Comment (felixfbecker) [inline index.html]: Indentation is inconsistent here

File: index.html (hunk 3 above)
Comment (ggilmore) [reply index.html]: @felixfbecker  If you look in the above screenshot, this box doesn't have a check in it.

**Comment by ggilmore** [comment]
@felixfbecker PTAL

screen shot 2018-05-09 at 12 55 03-fullpage

**Comment by laughedelic** [comment]
IMO it looks now more like an advertisement of Sourcegraph than an information page about LSP, because instead of completing the LSP features list, you add these extensions that are supported only by 5 servers (yes, it's a handful, but 4 of those are affiliated with Sourcegraph). 
I understand it's maintained by Sourcegraph so it's completely reasonable, but I think this bias should be clarified for the visitors of the website.

**Comment by ggilmore** [comment]
Hi @laughedelic,

This isn't intended to be an ad. It would certainly not be an effective one. 🙂 You are right that the table doesn't convey the extension info in the right way from the POV of users and implementors. We're going to condense the supported extensions into a single column or something similar.

**Comment by laughedelic** [comment]
@ggilmore thanks for reply. I don't have anything against non-LSP extensions and I think it's useful to know that some language servers have those capabilities and I like the way it looks now in #93! 👍

**Pull request merged**